File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.cc
#include <iostream>
#include <vector>

using namespace std;

int N; 
int num[11]; // 수 리스트
int operatorCount[4]; // 각 연산자의 개수 [+,-,×,÷]
vector<int> ansList; // 결과값 리스트
int answer = 0;

void solve(int n) {
    if (n == N) {
        ansList.push_back(answer);
        return;
    }
    for (int i = 0; i < 4; ++i) {
        if (operatorCount[i] == 0) continue;
        int temp = answer;

        operatorCount[i]--;  // 사용

        switch (i) {
            case 0:
                answer += num[n]; break;
            case 1:
                answer -= num[n]; break;
            case 2:
                answer *= num[n]; break;
            case 3:
                answer /= num[n]; break;
        }

        solve(n + 1);

        // 백트래킹
        answer = temp;
        operatorCount[i]++;
    }
}

int main() {
    cin >> N; // 수의 개수 입력

    for (int i = 0; i < N; ++i) {
        cin >> num[i]; // 수 입력
    }

    for (int i = 0; i < 4; ++i) {
        cin >> operatorCount[i]; // 연산자 개수 입력
    }

    answer = num[0];
    solve(1); // 함수 호출

    // 결과 출력
    int min = ansList[0];
    int max = ansList[0];
    for (int i=1; i<ansList.size(); i++){
        if (ansList[i]>max) max = ansList[i];
        else if (ansList[i]<min) min = ansList[i];
    }
    cout << max << '\n' << min;

    return 0;
}

File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.py
def solve(n):
    global answer
    if n==0:
        answer = num[0]
        solve(1)
        return
    if n==N:
        ans_list.append(answer)
        return
    for i in range(4):
        if operator[i]>0:

            temp = answer # temp에 이전 값 저장
            operator[i] -= 1

            # 연산 실행
            if i==0:
                answer += num[n]
            elif i==1:
                answer -= num[n]
            elif i == 2:
                answer *= num[n]
            elif i == 3:
                if answer < 0:
                    answer = -answer
                    answer //= num[n]
                    answer = -answer
                else:
                    answer //= num[n]

            # 다음단계 실행
            solve(n+1)

            #다시 넣기 이전으로 복구
            answer = temp
            operator[i] += 1


import sys

N = int(sys.stdin.readline())
# N 개의 수 리스트
num = list(map(int, sys.stdin.readline().split()))
# 각 연산자의 개수 [+,-,×,÷]
operator = list(map(int, sys.stdin.readline().split()))

ans_list = []

solve(0)

print(int(max(ans_list)))
print(int(min(ans_list)))